src/modules/task/internal/task-writer.ts

```diff
@@ -0,0 +1,53 @@
+import {
+  CreateTaskParams,
+  DeleteTaskParams,
+  GetTaskParams,
+  Task,
+  TaskWithNameExistsError,
+} from '../types';
+import TaskRepository from './store/task-repository';
+import TaskReader from './task-reader';
+import TaskUtil from './task-util';
+
+export default class TaskWriter {
+  public static async createTask(params: CreateTaskParams): Promise<Task> {
+    const existingTask = await TaskRepository.taskDB.findOne({
+      account: params.accountId,
+      name: params.name,
+      active: true,
+    });
+    if (existingTask) {
+      throw new TaskWithNameExistsError(params.name);
+    }
+    const createdTask = await TaskRepository.taskDB.create({
+      account: params.accountId,
+      name: params.name,
+      active: true,
+    });
+    return TaskUtil.convertTaskDBToTask(createdTask);
+  }
+
+  public static async deleteTask(params: DeleteTaskParams): Promise<void> {
+    const taskParams: GetTaskParams = {
+      accountId: params.accountId,
+      taskId: params.taskId,
+    };
+    const task = await TaskReader.getTask(taskParams);
+    await TaskRepository.taskDB.findOneAndUpdate(
+      {
+        _id: task.id,
+      },
+      {
+        $set: {
+          active: false,
+        },
+      },
+      {
+        new: true,
+        upsert: true,
```

Comment:
For delete upsert should not be true. It will create a new record. Please take a look at [mongoose docs](https://mongoosejs.com/docs/api.html#model_Model.findOneAndUpdate).

Reply:
@akashlakhera  please check , we are performing findOneAndUpdate not findOneAndDElete here, we are just updating the active field to false.

Reply:
Why do we need upsert?

Reply:
yes this was redundant part of code, not needed already removed it.